### ISSUE TITLE
Be able to display the instantaneous speed of the loop

### ISSUE THREAD

**Issue by BertrandBordage** [tqdm/tqdm#48]
Currently, tqdm shows the average speed of the loop, given the time it took since it started and the amount of iterations done. In some common cases this is not helpful, and the instantaneous speed would be better (the instantaneous speed being the inverse of the amount of time since the last iteration).

The classic case where this would be useful is for stoppable scripts that skip what was already done when you run them a second time after stopping them.  For example, a script that converts a million images and checks before that the converted image already exists. If you already converted 20% of the images, the second run of the script will execute 20% of the iterations at 300 its/s, while the standard speed should be 1 it/s. When the script reaches the first unconverted image, it still says that the speed is around 300 its/s, and you’ll have to wait several thousands of iterations (so several hours) before the speed comes back to something more reasonable (but the estimated remaining time will still not be correct).

For such scripts, it could therefore be useful to display the instantaneous speed, or we could even create an algorithm that uses the minimum, maximum, and median speed per iteration to give a realistic estimation of speed and remaining time that would work for any use case, instead of using only the average **or** the instantaneous speed depending on a boolean setting. The new estimations will be much more dynamic, but not ridiculously changing like Windows progress bars that only use the instantaneous speed and can switch from a few seconds to several hours in a heartbeat.

More statistics than minimum, maximum, and median speed (like quartiles) would increase the precision of the estimation, but these three values are enough for now in my opinion.

**Comment by casperdcl** [comment]
The `gui` version is a bit like windows 8/9 file transfer progress bars...

I've thought about this quite a bit, and think the best solution might be to take the average speed of the last 10 iterations...

**Comment by lrq3000** [comment]
I agree, this is something I have thought about before, but I'm not sure what would be the best  estimate.

As @casperdcl said, a common method is to use a running average or a running median filter (which is equivalent to a low-pass filter), to avoid noisy iterations such as the 20% you mention.

Of course we can implement several different methods and let the user choose, but I think that if we're going to implement new methods, we might as well do some research about it as to find good methods to implement.

**Comment by lrq3000** [comment]
After a quick research, here are some results:

- Exponential weighted moving average: indeed it's very common in DSP. More infos [here](http://stackoverflow.com/a/3841706/1121352), [here](http://stackoverflow.com/a/2442084/1121352) and [here](http://stackoverflow.com/a/7674088/1121352).

- Using not only a window over the number of iterations but also a [window over the last x minutes](http://stackoverflow.com/a/2779645/1121352).

- Improving ergonomics, eg, by dynamically updating the min_interval variable depending on how much time it will take. Example [here](http://stackoverflow.com/a/1883169/1121352) and [research paper here](http://ux.stackexchange.com/a/54971/45053).

- Discussions on why it is so hard (in fact impossible) to correctly predict remaining time in all cases: [here](http://www.howtogeek.com/139118/htg-explains-why-are-progress-bars-generally-inaccurate/), [here](http://bramcohen.livejournal.com/24122.html), [here](http://ask-leo.com/why_are_progress_bars_so_awful.html) and [here](https://www.reddit.com/r/askscience/comments/1mke2o/why_doesnt_windows_estimated_time_remaining_ever/).

And the necessary XKCD sketch:

![XKCD sketch Estimation](http://imgs.xkcd.com/comics/estimation.png)

So in the end, there's no perfect way to tackle this issue. Having a good time predictor requires to use machine learning algorithms on a sample dataset, which we cannot have in our case since we want the bar to be as general as possible. I think we should implement the "Exponential weighted moving average", which seems to be the go-to way of implementing reliable progress bars nowadays.

In addition, we could implement some kind of memorization of the iterations' durations. We could then make a pessimistic progress bar that would give more weights to the slowest iterations, and less to the fast iterations,

**Comment by lrq3000** [comment]
And of course, we can implement the algorithms the OP suggested, and leave the choice to the user.

**Comment by BertrandBordage** [comment]
@lrq3000 I agree with you, the exponential weighted moving average is probably the best solution for us (for the moment) for these reasons:
- it has no impact on the memory usage and an insignificant impact on CPU performance
- the smoothing factor can be changed by users, depending on the stability of the iteration speed
- it can act as a boolean to determine whether we want average speed or instantaneous speed: for average speed, use 0 as the smoothing factor, otherwise use 1

I suggest that we name that factor `speed_smoothing`, and that we set it to something like 0.7 by default.

This could eventually be improved later by using real statistics, but it would already be far better with that simple system, so I’m not sure it would worth developing it :)

**Comment by casperdcl** [comment]
K. Gonna implement ema then.

**Comment by BertrandBordage** [comment]
Thank you very much :D

**Comment by casperdcl** [comment]
Ok just writing tests for this. I tried to reinstall some coverage-related packages on my system a few days ago and now my coverage is broken.... Any Ideas how to fix? Been driving me nuts:

```sh
$ sudo -H pip show codecov coverage nose

---
Name: codecov
Version: 1.6.2
Location: /usr/local/lib/python2.7/dist-packages
Requires: coverage, requests
---
Name: coverage
Version: 3.7.1
Location: /usr/local/lib/python2.7/dist-packages
Requires: 
---
Name: nose
Version: 1.3.7
Location: /usr/local/lib/python2.7/dist-packages
Requires: 

$ nosetests --with-coverage --cover-package=tqdm -v tqdm/

tests_tqdm.test_format_interval ... ok
...
tests_version.test_version ... ok

Name            Stmts   Miss  Cover   Missing
---------------------------------------------
tqdm/__init__       6      6     0%   1-7
tqdm/_tqdm        177    177     0%   12-668
tqdm/_utils         6      6     0%   13-20
tqdm/_version       2      2     0%   2-5
---------------------------------------------
TOTAL             191    191     0%   
----------------------------------------------------------------------
Ran 17 tests in 0.006s

OK
```

**Comment by kmike** [comment]
@casperdcl likely tests are executed against installed tqdm, but coverage checks lines of tqdm in your folder. `tox

**Comment by casperdcl** [comment]
nvm. solved. (in case you're interested: `sudo pip uninstall nose` still left a bunch of `*.so` files under `/usr/local/lib/python2.7/dist-packages/nose` (not a full uninstall). Just had to clean out the folder and clean install. ;)

**Comment by lrq3000** [comment]
For reference in the future, two other algorithms:

- [Simple Moving Median](https://en.wikipedia.org/wiki/Moving_average#Moving_median), equivalent to a low-pass filter or a median filter. The big advantage is that this algorithm is not susceptible to outliers (ie, big sudden and rare changes in the speed, but it adapts the rate quickly if the trend is regular).

- [Moving average regression model](https://en.wikipedia.org/wiki/Moving-average_model), the killer weapon to model the trend using autocorrelative machine learning algorithm :p This is not so hard to implement, but it will surely take more computing time than the current algorithm we use. But surely this would be the most precise time estimation we can get.
